R/dependecies.R

```diff
@@ -0,0 +1,437 @@
+      package_version(as.character(locked_package$version)) <
+      package_version(as.character(locked_package$latest_version_in_lockbox)))) {
+      locked_package$version <- locked_package$latest_version_in_lockbox
+  }
+  if (!is.na(locked_package$version) && exists_in_lockbox(locked_package)) {
+    dependencies <- dependencies_from_description(locked_package
+      , description_file_for(locked_package$name
+        , dirname(lockbox_package_path(locked_package))))
+  } else {
+    cat(crayon::blue("."))
+    output <- tryCatch(get_remote_dependencies(package), error = function(e) e)
+    if (is(output, "error")) {
+      stop(crayon::red(paste0("Dependencies could not be resolved for package: "
+        , package$name, " version: ", package$version
+        , " remote:", package$remote)))
+    } else {
+      package <- output$package
+      dependencies <- output$dependencies
+    }
+  }
+  dependencies <- strip_available_dependencies(dependencies)
+  dependencies <- strip_duplicate_dependencies(dependencies)
+  dependencies <- lapply(dependencies, add_latest_version_in_lockbox)
+  dependencies <- lapply(dependencies, function(dep) {
+    dep$parent_package <- package$name
+    dep})
+  dependencies <- lapply(dependencies, replace_with_lock, lock)
+  list(package = package, dependencies = dependencies)
+}
+
+strip_available_dependencies <- function(dependencies) {
+  dependencies <- lapply(dependencies
+    , function(package) {
+      if(package$name %in% pesky_namespaces) NULL
+      else package
+    })
+ dependencies[!vapply(dependencies, is.null, logical(1))]
+}
+
+#' We don't trust package authors to only put a package in once
+strip_duplicate_dependencies <- function(dependencies) {
+  dependencies[!duplicated(vapply(dependencies, function(d) d$name
+    , character(1)))]
+}
+
+#' Get the dependencies for a given package
+get_remote_dependencies <- function(package) {
+  UseMethod("get_remote_dependencies")
+}
+
+#' If a package is local we just read from the directory given
+get_remote_dependencies.local <- function(package) {
+  description_name <- paste0(package$dir, .Platform$file.sep, "DESCRIPTION")
+  list(package = package
+    , dependencies =
+      dependencies_from_description(package, read.dcf(description_name)))
+}
+
+#' For packages on CRAN we will extract to a temporary directory when we
+#' download the accurate remote DESCRIPTION file. Because these are tarballs
+#' there is no simple way to extract only our desired file like we can with
+#' zipfiles using the unz function.
+get_remote_dependencies.CRAN <- function(package) {
+  original_version <- package$version
+  if (package$is_dependency_package) {
+    package$version <- NA
+  }
+  filepath <- download_package(package)
+
+  sep <- .Platform$file.sep
+  split_fp <- strsplit(filepath,sep)[[1]]
+  dirpath <- dirname(filepath)
+  file_list <- untar(filepath, list = TRUE)
+  description_name <- file_list[grepl(paste0("^[^", sep, "]+", sep
```

Comment:
Surely no author is evil enough to include a `.DESCRIPTION` file or some nonsense :)